master/src/main/java/org/apache/celeborn/service/deploy/master/SlotsAllocator.java

```diff
@@ -269,6 +278,51 @@ private static StorageInfo getStorageInfo(
     return slots;
   }
 
+  /**
+   * The rack distribution of the input workers list is essentially random, and in degenerate cases
+   * the rack aware slot selection ends up skipping sub list's of hosts (in same rack) - which
+   * results in uneven distribution of replica selection. For example given worker list: [h1r1,
+   * h2r1, h3r1, h4r2, h5r2, h6r2] if primary is h3r1 and replica index is pointing to h1r1, it will
+   * skip both h2r1 and h3r1 in order to pick h4r2; and for the next slot, primary will be h4r2 and
+   * will skip all the r2 hosts in order to pick h1r1. This ends up being suboptimal where some
+   * hosts are picked a lot more than others (due to the worker and worker/rack distribution). In
+   * order to mitigate this, we reorder the worker list by redistributing the workers based on rack
+   * to increase the rack diversity between adjoining workers, so that we minimize skipping over
+   * consecutive hosts.
+   */
+  static List<WorkerInfo> generateRackAwareWorkers(List<WorkerInfo> workers) {
+
+    List<Map.Entry<String, LinkedList<WorkerInfo>>> sortedRackToHosts;
+    {
+      Map<String, LinkedList<WorkerInfo>> map = new HashMap<>();
+      for (WorkerInfo worker : workers) {
+        map.computeIfAbsent(worker.networkLocation(), key -> new LinkedList<>()).add(worker);
+      }
+      sortedRackToHosts = new ArrayList<>(map.entrySet());
+      // reverse sort by number of hosts per rack
+      sortedRackToHosts.sort(
+          (o1, o2) -> Integer.compare(o2.getValue().size(), o1.getValue().size()));
+    }
+
+    ArrayList<WorkerInfo> result = new ArrayList<>(workers.size());
+    int count = 0;
+    final int numWorkers = workers.size();
+    while (count < numWorkers) {
+      Iterator<Map.Entry<String, LinkedList<WorkerInfo>>> iter = sortedRackToHosts.iterator();
+      while (iter.hasNext()) {
+        LinkedList<WorkerInfo> workerList = iter.next().getValue();
+        result.add(workerList.removeFirst());
+        count++;
+        if (workerList.isEmpty()) {
+          iter.remove();
+        }
+      }
+    }
+    assert (sortedRackToHosts.isEmpty());
```

Comment:
Why does empty `sortedRackToHosts` need to assert?

Reply:
Let me drop it - this will not happen anymore (the loop ensures it)